File: queries/h13.R
tpc_h13 <- function(input_func, collect_func = dplyr::collect) {
  c_orders <- input_func("customer") %>%
    left_join(
      input_func("orders") %>%
        filter(!grepl("special.*?requests", o_comment)),
      by = c("c_custkey" = "o_custkey")
    ) %>%
    group_by(c_custkey) %>%
    summarise(
      c_count = sum(!is.na(o_orderkey))
    )

  c_orders %>%
    group_by(c_count) %>%
    summarise(custdist = n()) %>%
    ungroup() %>%
    arrange(desc(custdist), desc(c_count)) %>%
    collect_func()
}

File: ibis_tpc/h13.py
import ibis


def tpc_h13(con, WORD1="special", WORD2="requests"):
    """Customer Distribution Query (Q13)

    This query seeks relationships between customers and the size of their
    orders."""

    customer = con.table("customer")
    orders = con.table("orders")
    innerq = customer
    innerq = innerq.left_join(
        orders,
        (customer.c_custkey == orders.o_custkey)
        & ~orders.o_comment.like(f"%{WORD1}%{WORD2}%"),
    )
    innergq = innerq.group_by([innerq.c_custkey])
    innerq = innergq.aggregate(c_count=innerq.o_orderkey.count())

    gq = innerq.group_by([innerq.c_count])
    q = gq.aggregate(custdist=innerq.count())

    q = q.order_by([ibis.desc(q.custdist), ibis.desc(q.c_count)])
    return q
